### ISSUE TITLE
Type context not propagated for nested TypedDicts and lists

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#1994]
Originally reported here: https://news.ycombinator.com/item?id=46299570

ty emits an `invalid-assignment` *and* a `invalid-argument-type` diagnostic for this snippet.

I didn't have time to write a proper MRE yet, but it looks like we're not propagating the type context correctly at some point here?

```py
from anthropic.types import MessageParam

data: list[MessageParam] = [{"role": "user", "content": [{"type": "text", "text": ""}]}]
```

<details><summary>Full diagnostic output</summary>
<p>

```
error[invalid-assignment]: Object of type `list[MessageParam | dict[Unknown | str, Unknown | str | list[Unknown | dict[Unknown | str, Unknown | str]]]]` is not assignable to `list[MessageParam]`
 --> main.py:3:7
  |
1 | from anthropic.types import MessageParam
2 |
3 | data: list[MessageParam] = [{"role": "user", "content": [{"type": "text", "text": ""}]}]
  |       ------------------   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Incompatible value of type `list[MessageParam | dict[Unknown | str, Unknown | str | list[Unknown | dict[Unknown | str, Unknown | str]]]]`
  |       |
  |       Declared type
  |
info: rule `invalid-assignment` is enabled by default

error[invalid-argument-type]: Invalid argument to key "content" with declared type `str | Iterable[TextBlockParam | ImageBlockParam | DocumentBlockParam | ... omitted 13 union elements]` on TypedDict `MessageParam`
 --> main.py:3:29
  |
1 | from anthropic.types import MessageParam
2 |
3 | data: list[MessageParam] = [{"role": "user", "content": [{"type": "text", "text": ""}]}]
  |                             ----------------------------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^-
  |                             |                |          |
  |                             |                |          value of type `list[Unknown | dict[Unknown | str, Unknown | str]]`
  |                             |                key has declared type `str | Iterable[TextBlockParam | ImageBlockParam | DocumentBlockParam | ... omitted 13 union elements]`
  |                             TypedDict `MessageParam`
  |
info: Item declaration
  --> /home/shark/ty-reproducer/.venv/lib/python3.13/site-packages/anthropic/types/message_param.py:24:5
   |
23 |   class MessageParam(TypedDict, total=False):
24 | /     content: Required[
25 | |         Union[
26 | |             str,
27 | |             Iterable[
28 | |                 Union[
29 | |                     TextBlockParam,
30 | |                     ImageBlockParam,
31 | |                     DocumentBlockParam,
32 | |                     SearchResultBlockParam,
33 | |                     ThinkingBlockParam,
34 | |                     RedactedThinkingBlockParam,
35 | |                     ToolUseBlockParam,
36 | |                     ToolResultBlockParam,
37 | |                     ServerToolUseBlockParam,
38 | |                     WebSearchToolResultBlockParam,
39 | |                     ContentBlock,
40 | |                 ]
41 | |             ],
42 | |         ]
43 | |     ]
   | |_____- Item declared here
44 |
45 |       role: Required[Literal["user", "assistant"]]
   |
info: rule `invalid-argument-type` is enabled by default

Found 2 diagnostics

````

</p>
</details>

**Comment by ibraheemdev** [comment]
At first glance, this looks like a combination of https://github.com/astral-sh/ty/issues/1576 as well as the union annotation not being handled well -- we'd likely need multi-inference here, or typed dict tagged unions.

**Comment by cscanlin** [comment]
Maybe related:
```py
from typing import NotRequired, TypedDict


class Args(TypedDict):
    a: str
    b: bool
    c: NotRequired[str]


args: Args = {'a': 'hello', 'b': True}
args2: Args = {**args, 'c': 'world'}
```

```
error[missing-typed-dict-key]: Missing required key 'a' in TypedDict `Args` constructor
  --> test.py:11:15
   |
10 | args: Args = {'a': 'hello', 'b': True}
11 | args2: Args = {**args, 'c': 'world'}
   |               ^^^^^^^^^^^^^^^^^^^^^^
   |
info: rule `missing-typed-dict-key` is enabled by default

error[missing-typed-dict-key]: Missing required key 'b' in TypedDict `Args` constructor
  --> test.py:11:15
   |
10 | args: Args = {'a': 'hello', 'b': True}
11 | args2: Args = {**args, 'c': 'world'}
   |               ^^^^^^^^^^^^^^^^^^^^^^
   |
info: rule `missing-typed-dict-key` is enabled by default

Found 2 diagnostics
```

**Comment by ibraheemdev** [comment]
@cscanlin that looks related to https://github.com/astral-sh/ty/issues/1332. We don't correctly handle kwarg splats in dictionary literals, and likely need special-casing for typed dicts as well.

**Comment by ibraheemdev** [comment]
After https://github.com/astral-sh/ruff/pull/21930 lands, the remaining issue here is https://github.com/astral-sh/ty/issues/2265.
